source/ada/lsp-ada_handlers-refactor_imports_commands.adb

```diff
@@ -227,68 +233,67 @@ package body LSP.Ada_Handlers.Refactor_Imports_Commands is
-              (LSP.Messages.TextEdit (Edit));
+            if Edits.changes.Contains (Self.Where.textDocument.uri) then
+               Edits.changes (Self.Where.textDocument.uri).Append
+                 (LSP.Messages.TextEdit (Edit));
+            else
+               declare
+                  Text_Edits : LSP.Messages.TextEdit_Vector;
+               begin
+                  Text_Edits.Append (LSP.Messages.TextEdit (Edit));
+                  Edits.changes.Include
+                    (Self.Where.textDocument.uri, Text_Edits);
+               end;
+            end if;
          end if;
       end if;
 
       --  Add with clause.
 
       if not LSP.Types.Is_Empty (Self.With_Clause) then
          declare
-            Withed_List : constant Libadalang.Analysis.Ada_Node
-              := Node.Unit.Root.Children (1);
-            Already_Imported : Boolean := False;
-            use type LSP.Types.LSP_String;
+            Last : Boolean;
+            S    : constant Libadalang.Slocs.Source_Location :=
+              Laltools.Common.Get_Insert_With_Location
+                (Node      => Laltools.Common.Get_Compilation_Unit (Node),
+                 Pack_Name =>
+                   Langkit_Support.Text.From_UTF8
+                     (LSP.Types.To_UTF_8_String (Self.With_Clause)),
+                 Last      => Last);
          begin
-            Loc := LSP.Lal_Utils.Get_Node_Location (Withed_List);
-
-            if Node.Unit.Root.Children (1).Children'Length = 0 then
-               --  If this compilation unit does not have 'with' clauses, then
-               --  simply add the 'with' statement.
-
-               Edit.span := (Loc.span.first,
-                             Loc.span.first);
-               Edit.newText := LSP.Types.To_LSP_String
-                 (LSP.Types.To_UTF_8_String ("with " & Self.With_Clause & ";")
-                  & ASCII.LF);
-            else
-               --  Otherwise, check if it already imports the seggested unit,
-               --  and if not, add the 'with' statement.
-
-               Edit.span := (Loc.span.last,
-                             Loc.span.last);
-
-               Already_Imported_Loop :
-               for With_Clause of Withed_List.Children loop
-                  if Self.With_Clause = LSP.Types.To_LSP_String
-                    (Langkit_Support.Text.To_UTF8
-                       (With_Clause.As_With_Clause.F_Packages.Text))
-                  then
-                     Already_Imported := True;
-                     exit Already_Imported_Loop;
-                  end if;
-               end loop Already_Imported_Loop;
-
-               if not Already_Imported then
-
+            if S /= Libadalang.Slocs.No_Source_Location then
+               Edit.span := LSP.Lal_Utils.To_Span (S);
+               if Last then
                   Edit.newText := LSP.Types.To_LSP_String
-                    (ASCII.LF & LSP.Types.To_UTF_8_String
+                    ("" & ASCII.LF
```

Comment:
it's not clear to me why do we need `"" &`? Looks like an extra do-nothing-operation.